### ISSUE TITLE
destinations of motorway junctions are not  shown or are shown too late

### ISSUE THREAD

**Issue by Neuwessi11** [osmandapp/OsmAnd#11778]
Hi!

I would like to re-open issue #8767 and have already written a comment there. Should I write this comment here again?
I am not quite sure which is the correct way, sorry for that.

**Comment by scaidermern** [comment]
I fully agree. I already tried to explain this in #9592 but it looks like it wasn't clear enough for the developers to understand/fix this.

With the current behavior I would even go so far to say that OsmAnd is useless(!) on (German?) motorway junctions. The text instructions and the voice instructions don't match the signs.

Example from the motorway junction "Kreuz Nürnberg Ost" from 49.393477,11.1886096 to 49.3990072,11.1982441:
OsmAnd shows these instructions:
![photo_2021-05-27_08-39-45](https://user-images.githubusercontent.com/1040784/119777910-1c72ff80-bec7-11eb-825c-9ba7a185e252.jpg)
The text and voice instructions just contain the number (60) of the motorway junction (https://www.openstreetmap.org/node/21298202). However this _doesn't_ help when trying to find the correct exit, because the motorway signs don't display this exit number! The motorway signs display the destinations, but OsmAnd does neither show nor announce these destinations.

See the following mapillary image where this exit number is _not_ contained: https://www.mapillary.com/app/?lat=49.393930863936916&lng=11.191060353694528&z=17&focus=photo&pKey=X0mEfka7NGAYTq8bfIplAA
![](https://images.mapillary.com/X0mEfka7NGAYTq8bfIplAA/thumb-640.jpg)

Instead of showing the exit _number_ OsmAnd should display the exit _destinations_ instead. This is already done for regular motorway exits and other junctions and works really well. I think OsmAnd also did this in older releases before showing the motorway exit number, so this is actually a regression.

Currently the only way to find the correct exist is to look at the turn lane instructions (see the screen shot). The text and voice instructions don't help at all.

This should get fixed in the next release because it is really a major limitation.

**Comment by vshcherb** [comment]
It was discussed in another topics with lots example in Germany (photos), Exit number already present on some autobahns and they will be updated more & more to be widespreaded further. It's possible to disable exit number announcement but anyway it becomes a worldwide standard for complex junctions (for example NL has it everywhere). Cause this is the only way (except Destination name) to have displayed it correctly.

It's correct that City Destination is 2nd best option to choose, unfortunately the Mapping in OSM is not clear (as far as I know), mapping "destination" should happen after intersection, so it's possible for routing to detect which road corresponds to which destination.

3rd option which is used and displayed is "ref" but in this case it's also not present and it is unlikely to be widespreaded in Germany vs other countries.

**Comment by vshcherb** [comment]
Why  "Kreuz Nürnberg Ost" was displayed, I don't know, it looks very odd (cc @sonora). I don't think junction name should be displayed at all.

**Comment by scaidermern** [comment]
> Exit number already present on some autobahns and they will be updated more & more to be widespreaded further. It's possible to disable exit number announcement but anyway it becomes a worldwide standard for complex junctions (for example NL has it everywhere).

In that case it will be nice to have them shown. But at the moment they aren't present at many (if not most) German motorway junctions. And even if they will be present at some day in the future I can't imagine that they will replace destinations. So destinations should probably get shown anyway.

> 3rd option which is used and displayed is "ref" but in this case it's also not present because it's tagged as "destination:ref"? That doesn't look very common in OSM.

I'm not sure about other parts of the world but `destination:ref` is pretty common around here:
- https://overpass-turbo.eu/s/17MN
- https://overpass-turbo.eu/s/17MO

**Comment by vshcherb** [comment]
https://github.com/osmandapp/OsmAnd/issues/9592#issuecomment-703312162

**Comment by scaidermern** [comment]
> [#9592 (comment)](https://github.com/osmandapp/OsmAnd/issues/9592#issuecomment-703312162)

Right.

The first one is a large sign that displays the exit number. But the sign doesn't say which exit to take. So it is not really useful for the driver.

The second one is a pretty small sign that exists just right before the motorway exit. It is hard to see and it is already too late for the driver to properly react upon seeing it. So also not very helpful.

Showing the exit number only makes sense if it is also present on these large signs that also tell you which exit / which lane to take (see the image at https://github.com/osmandapp/OsmAnd/issues/11778#issuecomment-849379687).

**Comment by vshcherb** [comment]
@scaidermern I wouldn't argue much about Germany since even different lands could have different prospects on infrastructure. Though OsmAnd should work worldwide and neighboor NL have great use of this Exits. 

Anyway I would keep this bug open only for discussion:
- what should be displayed in name (why destination is missing and what exactly)
- why ref is not displayed (destination:ref is not supported?)

For Exit number display or not please open another discussion.

**Comment by Neuwessi11** [comment]
@vshcherb 
Why should scaidermern not argue much about Germany? It is a not very small country in the middle of Europe, where millions of people from whole of Europe (from Russia to Portugal as well as from North-Norway to south Italy) are driving. And it is not only about Germany, the problem is exactly the same for instance in Sweden. About Denmark I will have a special look in the beginning of July.



> Why "Kreuz Nürnberg Ost" was displayed, I don't know, it looks very odd (cc @sonora).

That I can tell you: Because that is the way OsmAnd works. "Kreuz Nürnberg Ost" is the junction name, completely correctly tagged in OSM. And because the junction name is preferred before the destinations of the splitted road, it is shown right until the end.
And because destinations are only supported when passing the splitting point (what is more or less the point of no return :) ) there is no time to prepare for the correct lane.
I can only use the direction arrows. But these are quite small and not always correctly shown.

I must absolutly agree to scaidermern: It is a really major limitation.

About the exit numbers: My impression is, that is not the topic. It is good to have them, who is annoyed by them can disable them. Everything fine. 

Again: The major point are the missing destinations.

**Comment by scaidermern** [comment]
> Again: The major point are the missing destinations.

Exactly.

I don't care about the exit number. Displaying/announcing it is completely fine. But it should not replace displaying/announcing the destinations.

Also of course I'm interested in a solution that works world-wide. I've taken examples from Germany because that's the region I'm usually around. I can't tell much about motorways and signs from other countries. But I'm pretty sure that destinations also play an important role in other countries.

**Comment by sonora** [comment]
For Germany, let's look at the example signs depicted on this page http://www.strassenschilder.de/richtzeichen/vorwegweiser-auf-autobahnen/:

**Exit numbers**  (the "26" in white **round**, not cornered circle) are shown on German Autobahns (for **exits**, highway junctions follow a different pattern starting at 2000 m!):
 - on the 1000 m sign (see the "_Düsseldorf-Benrath_" example)
 - and again on top of the 300 m sign with the 3 white bars (no example shown on that web site).
 - They are **not** shown on the 500m sign (on this web site the "_Dietz/Wallmerod_" example)

Regarding text (**for exits**)
- the 1000 m sign shows an exit "name" (see example "_Düsseldorf-Benrath_"), under which the exit is known on all conventional maps, and this usually coincides with the prominent/first "destination" an exit leads to.
- the 500 m sign shows all "destinations" ("_Dietz_ and _Wallmerod_"), plus the name of the next after that.
- the immediate sign again shows all "destinations" (example here "_Mainz_ / _Wiesbaden_"), is not always used like this

I can confirm that our logic works fine for this exit use case:
- At about ~1.5km away we display (and announce) both the **exit number**, and **name** of the exit to take. For the case of **motorway exits** I find no example where this is a problem, because usually whatever we use as "name" here is what the exit is known under on all conventional maps, and this usually coincides with the prominent/first "destination" an exit leads to. (In the short announcement we then finally announce **all destinations**.)

- The actual issue appears for highway **junctions**: We treat these exactly like exits, meaning further out we show / announce their "name", only in the final announcement (when you are near the turn lanes), we announce the destinations. The "name" for the "true exit case" coincides with a destination, so no issue there.
  **But in this intersection case the "name" of the highway intersection is just its name, which usually specifies its location but contains zero directional information.**

I am not sure if it has always been like this, it feels like a regression to me: I think I should have noticed on many occasions in years past. Possible that this was introduced when we did the work supporting exit numbers. Also, I think this may be an issue in lots of countries, not just in Germany.

The possible fix could be to replace whatever I call "name" in the above by actually a destination. With the side effect that many exits (and intersection junctions) have a longer list of destinations, so it would mean long displays and repeated long audio prompts. Or we somehow for the further-out prompt just shorten to the first destination. (In most countries this is farthest, but in some the closest.)

**Comment by Neuwessi11** [comment]
@sonora 
I must contradict you in some points.

>    For the case of **motorway exits** I find no example where this is a problem,

Maybe for you, for me it is a real problem. Sorry ...

>    * because usually whatever we use as "name" here is what the exit is known under on all conventional maps, and this usually coincides with the prominent/first "destination" an exit leads to. 

The problem here is the word "usually". What in cases where not? And you can regularly find signs, where are slight differences. These mostly in urban regions, where the exits follow sometimes in the distance of 200 or 400 m. In my opinion in such a situation is everything, that is not like the real life, unsettling and possibly leading to mistakes.

>    * (In the short announcement we then finally announce **all destinations**.)

No, normally not.
Sometimes I can get a one-second-glance, but everytimes AFTER passing the splitting point. What is too late.


>     * only in the final announcement (when you are near the turn lanes), we announce the destinations.

I am really sorry, but this is not true. The destinations are announced at the moment where the turn lanes start.

>     * The "name" for the "true exit case" coincides with a destination, so no issue there.

I unfortunately don't understand what you mean with "true exit case".

>     * But in this intersection case the "name" of the highway intersection is just its name, which usually specifies its location but contains zero directional information.

What is a big problem, indeed. Should be fixed. When driving on such an intersection I am absolutely not interested in any location information. I need directional information. 

>  Also, I think this may be an issue in lots of countries, not just in Germany.

Yes, I can confirm that. Driving through Sweden I found a lot of places, where the exit name has either nothing to do with the destinations written on the exit sign, or is the last of three or four cities.

> The possible fix could be to replace whatever I call "name" in the above by actually a destination.

Why that? In Germany we would call this "throw out the child with the bath".
The intersection or exit name should be shown where it is shown at the moment, that is really fine, because I see this at this time. But either after passing the 1000 m sign or at least ~500 m ahead of the exit it should switch to the destinations.

Maybe for the audio it can be shortened to two destinations (as a compromise :), but for the display I would leave it originally.
Maybe at some time OsmAnd will show real-life-signs, and then it would be helpful to have the whole list.

**Comment by leogott** [comment]
I authored Issue #10718 where I pointed out a few related problems. I'll summarize them here:

When navigating the German Autobahn, OSMAnd's voice announcements generally aren't sufficient, because they often don't include the information relevant at the time, or include information that would have been very useful to have before passing the signs that were referred to.

image: Junction/Exit Number on a seperate sign
<img width="125" src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/7a/Zeichen_406-50_-_Knotenpunkte_der_Autobahnen_%28ein-_oder_zweistellige_Nummer%29%2C_650x650%3B_StVO_2013.svg/240px-Zeichen_406-50_-_Knotenpunkte_der_Autobahnen_%28ein-_oder_zweistellige_Nummer%29%2C_650x650%3B_StVO_2013.svg.png"/>

1) On many but not all parts of the Autobahn, Highway exits have a number, which would be immensely useful to have displayed and announced at the time of the "stay on this road for distance" message. e.g. "Drive on the A1 for 50km until [junction name] Exit number 99".
The numbers can be relevant again, if present on the signs, when approaching the junction or exit, to correctly identify which exits **_not_** to take.

Currently the exit numbers are usually announced _after_ the sign "exit number soandso in x km" was passed, which is not useful.

2) Which lanes to take when approaching a junction often is _very_ important, and I've missed my share of exits because I couldn't look at the display at the time.

Compare this mapillary view: https://www.mapillary.com/map/im/l5iLCn2FftLp1j1MHUtthE (not sure if/how to embed)
Shown is the latest possible moment to take the parting lanes. Ideally OSMAnd would have already informed the driver of the Destination to look for, and instructed them to take one of the two rightmost lanes half a minute ago. This is not currently the case. The [take turn now] message would probably be playing right now, and the destination to look for would be in the next turn message, and not even on screen.

3) Semi-related problem: Turns often are scheduled for the latest moment they could be taken, but when an exit starts its own lane, there are official signs
<img width="125" src="https://upload.wikimedia.org/wikipedia/commons/thumb/7/71/Zeichen_450-51_-_Ank%C3%BCndigungsbake_zweistreifig_%28200_m%29%2C_StVO_2013.svg/260px-Zeichen_450-51_-_Ank%C3%BCndigungsbake_zweistreifig_%28200_m%29%2C_StVO_2013.svg.png">
counting down to an earlier point where the turn can be taken, and OSMAnd should use the same point, as the status quo is super confusing and again - leads to missed turns.


**tl;dr:**
Navigating the German Autobahn would be much easier if for more complex junctions (most of them) PREPARE_TURN messages included the direction of the **after next** TAKE_TURN message. (Because to the driver, it's not two turns after another, it's one complex turn).
Exit/Junction Numbers can be immensely helpful, but they need to be announced earlier, and one junction can have like five exits, which _aren't numbered individually._

**Comment by scaidermern** [comment]
> **Exit numbers** (the "26" in white **round**, not cornered circle) are shown on German Autobahns (for **exits**, highway junctions follow a different pattern starting at 2000 m!):
> 
>     * on the 1000 m sign (see the "_Düsseldorf-Benrath_" example)
> 
>     * and again on top of the 300 m sign with the 3 white bars (no example shown on that web site).
> 
>     * They are **not** shown on the 500m sign (on this web site the "_Dietz/Wallmerod_" example)
> 

There isn't a single sign showing the exit number in my previous example ("Kreuz Nürnberg Ost"). And this is really just a random example, this also applies to various other motorway junctions.

**Comment by Neuwessi11** [comment]
I observed a similar and maybe connected problem sometimes ago, but today I confirmed it on two spots here at home.
It doesn't matter if it's in town or not, when this sign is visible

![grafik](https://user-images.githubusercontent.com/70072237/120078081-f8254780-c0ad-11eb-8b7b-23012733484c.png)

OsmAnd only shows "A 2", what is in most cases useless, because there are two ways leading to the A 2, but in different directions.
The destination (in that case "Berlin") is shown by OsmAnd only when passing the sign, what is too late, because for orientation I must see it depending on the maxspeed of the street about 100-300 m ahead.

I am not sure if this is already topic in another issue, but I thougt it might be similar to the topic in this issue.

**Comment by sonora** [comment]
Let me summarize again the ground truth for Germany, so I can get a picture of which situations you have he issues with:

**Numbers:**
- For **exits** exit  numbers are displayed on the 1000m and the 300 m sign
- For **junction exits** exit numbers are displayed on the 2000 m and the 300 m signs.

**Text:**
- For **exits**: 1000 m sign shows exit name only, 500 m sign shows destinations, turn lane sign shows destinations
- For **junction exits** 2000 m shows name (which bears no destination reference), 1000 m,, 500 m and turn lane signs all show destinations.

This is the situation for the vast majority of situations, or do I describe this wrong? (Of course there can be deviations in crammed settings. But we will have to find a navigation aid pattern which works independent of sign locations, because it is impossible to know all sign positions in all countries anyway.)

**Possible approaches:**
Distilling from the ideas above, I find the following suggestions:

- (a) "Continue" long haul message could be enhanced like "_Drive on the A1 for 50km until [junction name] Exit number 99_".
- (b) We may investigate never using any junction name, but always use destinations, where needed in abbreviated form (the effect of this needs to be evaluated internationally.)
- (c) We may have to include the junction number on all prompts, as its use and sign positioning is very inhomogeneous internationally

**Comment by vshcherb** [comment]
Few comments:
1. (c) - is already done by having a separate setting for this.
2. The issue might look simple but it is not in a core of it. It really depends on how tagging is done in different areas. Even detecting 'ref' + 'destination:ref' could be complex on multi segment turn. Though we need to take it example by example. 

So I think the nice example is given this https://github.com/osmandapp/OsmAnd/issues/11778#issuecomment-849379687 and we could start with it.

**Comment by scaidermern** [comment]
> @scaidermern Do you happen to be in that area, can you try driving the **opposite** direction coming from the east on A6 turning north on A3?

Unfortunately not. I'm only occasionally driving by car. In these cases usually rather long distances, that's why I experienced that this is currently broken for many motorway junctions.

**Comment by Neuwessi11** [comment]
Hi sonora,

> This is the situation for the vast majority of situations, or do I describe this wrong?

No, your description was very precise.

As to your ideas, (a) sounds quite nice to me, but again (b) seems to me sounding like overreacting. 
I have no idea in which way exits and junctions are signed in other countries (except Sweden :)), but why not first show the name (of exit and/or junction) and later show the destinations IF TAGGED?
Your American example seemed to me signed only with number and name. Then of course it should be displayed only the name and the number, but in cases where destinations are on the signs and are tagged in OSM, there it should be displayed.



> German exits I personally feel work rather sufficient, as do American exits, my other big testing use case.

As I stated, I cannot say anything about American exits, but as to the German exits I explicitly want to contradict you: No, it does not work sufficient! The destinations are strongly missing! (I am sorry, but this is a really strong feeling ...)

**Comment by sonora** [comment]
@Neuwessi11 thanks for commenting. Regarding your 'strong feeling', just to make sure: You are talking about 'regular' German exits, i.e. leaving the Autobahn, not junction exits (changing onto another Autobahn)?

**Comment by sonora** [comment]
PS: I think you are right. I found an example even for a 'regular' exit: This exit called Schwandorf-Mitte (coming south on the Autobahn A93) is a split exit, you may have to exit before the overpass of B85 (to go west) or after (to go east). But both are called the same, and without seeing/hearing direction infos it may be difficult, I have just run the simulations.

![20210603_090615](https://user-images.githubusercontent.com/845735/120602898-bf89c300-c44b-11eb-9253-129d9374b224.jpg)

**Comment by Neuwessi11** [comment]
@sonora 


> Regarding your 'strong feeling', just to make sure: You are talking about 'regular' German exits, i.e. leaving the Autobahn, not junction exits (changing onto another Autobahn)?

Yes, I was talking about regular exits.

But this does of course in no way affect the fact (in my eyes) that junctions are a problem, too. Maybe even the bigger problem.
With this 'strong feeling' I wanted to express that I really miss the destinations not only on junctions, but on exits as well.

**Comment by Neuwessi11** [comment]
@sonora 
> And I guess your suggestion is to simply **add** the (maybe just first) destination info after the name info?

No, that was a misunderstanding.

Because on the 2000m signs (regarding junctions) or the 1000m signs (regarding exits) always the name is written, OsmAnd should first show the name (of the junction/exit). 
But later the name is never signed again and is replaced by the destinations. That's why later (e.g. 1700m prior to a junction/700m prior to an exit, maybe even earlier) displaying the name should be REPLACED by displaying the destinations (if tagged). And because I think it best to have the display being closest to the real life, I woud suggest to display as many destinations as possible. 
I don't know if such a display scheme would be not as good in other countries, in Germany it would work absolutly perfect. 

And I want to add that because I always drive with disabled voice prompts, everything I say is always concerning the text and NOT the voice. About the voice I simply cannot say anything ...
But because of repeated very long voice prompts it sounds plausible to me to tell maybe only two destinations.
But again: As to voice I cannot say anything, and as to displaying destinations I would suggest to display as many as possible.

**Comment by sonora** [comment]
I drove our complex example 'Kreuz Altdorf' again the other day, coming from the east turning north. You have to leave A6 first onto a combined turn lane towards the intersecting A3, which then after 300m splits leading north or south on A3.

Could not watch the display but listened to the voice prompts which follow a similar logic. I received a total of four prompts:

(1) The very first one gave me destination info as posted on the signs, I think it said "_After about 2 km turn slightly right on exit 61 towards Regensburg, Wuerzburg, Nuernberg Nord_". So that one seems perfect.

(2) , (3)The second and third one just mentioned exit 61 '_Kreuz Altdorf_', with no destination info. The third one (it would actually be the final exit prompt for a non-complex exit) was at the beginning of the combined turn lane towards both intersecting destinations Regensburg (southbound) and Wuerzburg(northbound). That's also how the destinations are tagged in OSM on this part, but at this point we announce 'name'. You also hear suffix "_... then keep right_", correctly indicating the imminent split of the turn lane.

(4) Shortly after that the combined turn lane splits and you receive the correct destination information to "_turn right towards Wuerzburg,  Nuernberg Nord_".

So I guess the concern here in this complex situation  is anouncements (2) and (3) instrucring you to exit but stating only the overall intersection name, while the posted signs state desrinations.

I do not think there is much ambiguity in what OsmAnd told me, navigating this was ok. But I can see how in that case hearing/seeing destination info could  be even more useful.

**Comment by Neuwessi11** [comment]
Big thanks testing! 
Could you please make another try with disabled voice prompts?

as to 
> (2) , (3)The second and third one just mentioned exit 61 '_Kreuz Altdorf_', with no destination info. The third one (it would actually be the final exit prompt for a non-complex exit) was at the beginning of the combined turn lane towards both intersecting destinations

I think the reason for that is, that you wait with showing (or telling) the destinations until the driver passed the OSM-node 'motorway_junction'. This behaviour (waiting for showing the next one until passage of the first one) might fit to a sequence of destination signs, but not for the motorway_junction-point.


> I do not think there is much ambiguity in what OsmAnd told me

May be. But I think there should be NO ambiguity.
May be it has something to do with the difference between 'quantitiy' and 'quality' and is similar to genomes: The human genome is to 97,5% exactly the same like the genome of the mice (so much for 'quantity'), but (so much for quality) I think we agree that there is rather a huge difference between humans and mice. :)
Similar to OsmAnd: It is nice that it works to 97,5%, but if you are experiencing only one of the 2,5 remaining % of cases, it makes a real difference.

**Comment by Neuwessi11** [comment]
Sorry for annoying you, but during todays trip OsmAnd showed a rather weird behavior:
When driving the A 46 eastbound towards the Dreieck Wanlo (#10 of the A 46) with destination Mönchengladbach (which includes a change to the A 61), OsmAnd showed the destinations, but not the exit number and not the junction name. And I could reproduce it.
But driving the opposite route (coming A 61 from the north, and changing on Dreieck Wanlo to the A 46 direction Heinsberg), it showed the exit number and name, but no destinations. Is this a bug?

And I saw another thing: What I wrote on May 29, 18:49, worked today well on one spot. That means, OsmAnd seems to react inconsistent on different spots.

**Comment by sonora** [comment]
Possibly the difference is in the map data: Maybe the eastbound direction is mapped wirhout exit name and number?

**Comment by Neuwessi11** [comment]
No, I could not find any tagging difference: All directions of the motorway junction are tagged with exit number, exit name and destinations.

**Comment by vshcherb** [comment]
Short summary for first fix attempt (Example [1](https://github.com/osmandapp/OsmAnd/issues/11778#issuecomment-849379687) and [2](https://github.com/osmandapp/OsmAnd/issues/11852)).

Priority to show (especially on motorways, probably everywhere)
Slot 1: Junction ref

Slot 2:
1. Destination Ref
2. Ref

Slot 3:
1. Destinations
2. Name

**Comment by vshcherb** [comment]
With this partial fix name of exit name should not be shown.

**Comment by vshcherb** [comment]
![telegram-cloud-photo-size-2-5359628915347403411-y](https://user-images.githubusercontent.com/1042025/122593442-f83db500-d065-11eb-89e5-12f607aa1a8f.jpg)

**Comment by sonora** [comment]
(Have not tried this partial fix yet.) As an attempt for a complete fix, I guess we could simply swap the logic, meaning use what we now use at stage (1) (in my post further up) for stage (2) and (3), and use the current (2/3) logic (which uses the exit name) only for (1).

Only possible drawback may be that (2) and (3) will potentially become significantly longer.

**Comment by ivanPyrohivskyi** [comment]
https://github.com/osmandapp/OsmAnd/pull/12005

**Comment by ivanPyrohivskyi** [comment]
Tests:
turn slightly right onto exit 60 , toward A 9, Berlin Dresden, Bamberg

https://github.com/osmandapp/OsmAnd/issues/3289
from 48.0066, 11.6704 to 48.03167, 11.66855
in 750 meters, turn slightly right onto exit 95, toward Nürnberg, Stuttgart, Messe
turn slightly right onto exit 95, toward Nürnberg, Stuttgart, Messe
in 600 meters, turn slightly right toward A 99, Nürnberg, Stuttgart


https://github.com/osmandapp/OsmAnd/issues/3181
from 47.80970, 12.12140 to 47.80840, 12.11240
turn left onto Rosenheimer Strasse toward B 15, München, Innsbruck
turn left onto Rosenheimer Strasse toward B 15, München, Innsbruck, then in 40 meters, turn left
turn left on B 15 toward München, Innsbruck

**Comment by ivanPyrohivskyi** [comment]
https://www.openstreetmap.org/way/329734926
From 49.39387, 11.1907 to 49.40232, 11.19654

![Screenshot_1625145655](https://user-images.githubusercontent.com/1040293/124131353-6338b400-da88-11eb-8f50-02e094f04ebd.png)
![Screenshot_1625145658](https://user-images.githubusercontent.com/1040293/124131399-6fbd0c80-da88-11eb-98f5-a1ef9e1caf68.png)
![Screenshot_1625145674](https://user-images.githubusercontent.com/1040293/124131449-7e0b2880-da88-11eb-9b81-f550e9579d11.png)

TTS:
The trip is about 2 kilometers time is  one minute.
in 250 meters, turn slightly right onto exit 60 ,  toward A 9, Berlin Dresden, Bamberg
turn slightly right onto exit 60 ,  toward A 9, Berlin Dresden, Bamberg
in 300 meters, turn slightly right toward A 9, Berlin Dresden, Bamberg
turn slightly right 
in 300 meters, keep left toward A 9, Berlin, Bamberg
keep left 
Continue for 700 meters on A 9 and arrive at your destination 
You have reached your destination A 9

**Comment by ivanPyrohivskyi** [comment]
https://www.openstreetmap.org/way/145035982
From 48.0066, 11.6704 to 48.03157, 11.66834
From issue https://github.com/osmandapp/OsmAnd/issues/3289


![Screenshot_1625146529](https://user-images.githubusercontent.com/1040293/124133426-7d739180-da8a-11eb-846c-1207125f9ed4.png)
![Screenshot_1625146538](https://user-images.githubusercontent.com/1040293/124133509-911ef800-da8a-11eb-85a8-6c3ee47f3d15.png)
![Screenshot_1625146568](https://user-images.githubusercontent.com/1040293/124133573-a005aa80-da8a-11eb-9bc2-6abf73617328.png)


TTS:
The trip is about 3 kilometers time is  2 minutes. 
in 750 meters, turn slightly right onto exit 95 ,  toward Nürnberg, Stuttgart, Messe  , then keep right
turn slightly right onto exit 95 ,  toward Nürnberg, Stuttgart, Messe 
in 600 meters, turn slightly right toward A 99, Nürnberg, Stuttgart
turn slightly right 
Continue for about 1 kilometer to A 99 and arrive at your destination 
You have reached your destination A 99

**Comment by Neuwessi11** [comment]
That looks very nicely!

What does the app show/tell approaching the junction from 2.500 m?

**Comment by ivanPyrohivskyi** [comment]
http://www.openstreetmap.org/way/191799180
From issue https://github.com/osmandapp/OsmAnd/issues/3181
From 47.80970, 12.12140 to 47.80840, 12.11240

![Screenshot_1625147343](https://user-images.githubusercontent.com/1040293/124135309-5c13a500-da8c-11eb-833a-b718357a50c4.png)
![Screenshot_1625147356](https://user-images.githubusercontent.com/1040293/124135359-6897fd80-da8c-11eb-9c34-18d07a77ef1f.png)
![Screenshot_1625147363](https://user-images.githubusercontent.com/1040293/124135375-6c2b8480-da8c-11eb-9790-c3be320baaeb.png)


TTS:
The trip is about 2 kilometers time is  3 minutes. 
turn left onto Rosenheimer Strasse toward B 15, München, Innsbruck, then in 40 meters, turn left 
turn left on B 15 toward München, Innsbruck
Continue for about 2 kilometers on A 8 and arrive at your destination

**Comment by vshcherb** [comment]
We've tested following cases - https://github.com/osmandapp/OsmAnd/issues/11778#issuecomment-872246162 and think now situation is better . In case we have other ideas how to fix it, it's better to open separate issue and link it later to this one.

Comments / ideas / feedback could be commented here as well.

**Comment by sonora** [comment]
Quickly tested on the use cases we had collected here and it works like intended.

The prompts are a bit more 'repetetive' now, because in all instances we always repeat the destination lists, which can be lengthy in some places.(Instead of, as I suggested, 'swapping the logic', i.e. now in the first ('far out') case using the shorter exit name, and from the second prompt on use the dest list).

But let's see if we receive any negative feedback about that.
